engine/consensus/compliance/core.go

```diff
@@ -261,39 +298,69 @@ func (c *Core) processBlockAndDescendants(proposal *messages.BlockProposal, inRa
+		cpr := c.processBlockAndDescendants(childProposal, proposal.Header)
 		if cpr != nil {
 			// unexpected error: potentially corrupted internal state => abort processing and escalate error
 			return cpr
 		}
 	}
 
-	// drop all of the children that should have been processed now
+	// drop all the children that should have been processed now
 	c.pending.DropForParent(blockID)
 
 	return nil
 }
 
 // processBlockProposal processes the given block proposal. The proposal must connect to
 // the finalized state.
-func (c *Core) processBlockProposal(proposal *messages.BlockProposal, inRangeBlockResponse bool) error {
+// Expected errors during normal operations:
+//   - engine.OutdatedInputError if the block proposal is outdated (e.g. orphaned)
+//   - engine.InvalidInputError if the block proposal is invalid
+//   - engine.UnverifiableInputError if the proposal cannot be validated
+func (c *Core) processBlockProposal(proposal *messages.BlockProposal, parent *flow.Header) error {
 	startTime := time.Now()
-	defer c.complianceMetrics.BlockProposalDuration(time.Since(startTime))
+	defer func() {
+		c.hotstuffMetrics.BlockProcessingDuration(time.Since(startTime))
+	}()
+
+	header := proposal.Header
+	blockID := header.ID()
 
-	span, ctx, isSampled := c.tracer.StartBlockSpan(context.Background(), proposal.Header.ID(), trace.ConCompProcessBlockProposal)
+	span, ctx, isSampled := c.tracer.StartBlockSpan(context.Background(), blockID, trace.ConCompProcessBlockProposal)
 	if isSampled {
 		span.SetAttributes(
-			attribute.String("proposer", proposal.Header.ProposerID.String()),
+			attribute.String("proposer", header.ProposerID.String()),
 		)
 	}
 	defer span.End()
 
-	header := proposal.Header
+	hotstuffProposal := model.ProposalFromFlow(header)
+	err := c.validator.ValidateProposal(hotstuffProposal)
+	if err != nil {
+		if model.IsInvalidBlockError(err) {
+			return engine.NewInvalidInputErrorf("invalid block proposal: %w", err)
+		}
+		if errors.Is(err, model.ErrViewForUnknownEpoch) {
+			// We have received a proposal, but we don't know the epoch its view is within.
+			// We know:
+			//  - the parent of this block is valid and was appended to the state (ie. we knew the epoch for it)
+			//  - if we then see this for the child, one of two things must have happened:
+			//    1. the proposer malicious created the block for a view very far in the future (it's invalid)
+			//      -> in this case we can disregard the block
+			//    2. no blocks have been finalized within  the epoch commitment deadline, and the epoch ended
+			//       (breaking a critical assumption - see EpochCommitSafetyThreshold in protocol.Params for details)
+			//      -> in this case, the network has encountered a critical failure
+			//  - we assume in general that Case 2 will not happen, however we cannot prove Case 1, therefore we can discard this proposal
+			return engine.NewUnverifiableInputError("cannot validate proposal with view from unknown epoch: %w", err)
```

Comment:
```suggestion
			//  - we assume in general that Case 2 will not happen, therefore this must be Case 1 - an invalid block
			return engine.NewInvalidInputError("invalid proposal with view from unknown epoch: %w", err)
```

Since this case results in us marking the block invalid in the VoteAggregator, and we are assuming Case 1 (invalid block), suggesting we use InvalidInput error for consistency.